### DISCUSSION TITLE
Any plans for a WEB GUI ?

### DISCUSSION THREAD

**Discussion by shahidcodes** [0xERR0R/blocky#261] [Q&A]
Using blocky and really enjoying it. I was previously using PiHole which has a nice UI to inspect what's going to and stuffs. Just wanted to ask if there is any plan for any GUI?

**Comment by 0xERR0R** [comment]
Hey,

no, currently not. There is a Grafana dashboard for prometheus metrics, which shows some internal metrics and allows to enable/disable the blocking state.

I'm also currently working on (#258). This will allow to create another Grafana dashboard with detailed statistics about asked queries (e.g. top 20 asked domains, Top 20 blocked domains, last queries etc).

I'm trying to follow the KISS principle and implement only core features (basically DNS features) and provide integration for other tools (REST api, prometheus integration).

**Reply by shahidcodes** [reply]
That's great. If there's a way to communicate to blocky through api to fetch and manage things like upstream dns, clients, customDNS, client groups, block lists and so on then I think it'll be easier for me or anyone else to write a frontend.

At least for now I'm waiting for the WIP: logging query to database so that at least I can fetch the data from the database and see.

**Comment by 0xERR0R** [comment]
No, currently, you can only control the blocking (enable, disable), trigger the list refresh and perform a DNS query.

Here is the current REST API description: https://0xerr0r.github.io/blocky/swagger.html

**Comment by Mozart409** [comment]
Should I create a new frontend?

**Reply by 0xERR0R** [reply]
> Should I create a new frontend?

Do you mean a new frontend as replacement for the "simple" landing page or a full-fledged web UI (like pihole)? For the second: there are currently no plans for the web UI, but if we decide to make it, i would prefer to create a standalone web client which communicates per REST with blocky. In this case, core users can still use blocky without web UI and other can optionally install the web ui.

**Reply by Mozart409** [reply]
Yes something like pihole. A standalone would be a better option. What are the requirements for the frontend?

**Reply by 0xERR0R** [reply]
The problem is: blocky is at most stateless, configuration is in a file. There is currently no possibility to add / change a list entry, also a query log is not always persistent.

Do you have any ideas, which use cases should be implemented?

**Reply by Mozart409** [reply]
Yes. I have a demo project with all rest api endpoints. Its quick and dirty, but it works. [Project](https://github.com/Mozart409/blocky-frontend)

You can test the "production" version with this [docker-compose](https://github.com/Mozart409/blocky-frontend/blob/main/docker-compose.yml) file. Its included into the project. Be careful. The frontend runs on port [http://localhost:81](http://localhost:81)

Features

- enable / disable blocking
- Refresh Lists
- Test DNS Queries
- Responsive Design

**Reply by Mozart409** [reply]
I think blocky could allow for and api to mutate the file or e.g. fetch the list from redis and the frontend could mutate the datastore (e.g. redis, mysql, mongo) and refresh the list.

**Reply by Mozart409** [reply]
Blocky could maybe provide more info when checking dns queries. Blocky outputs the ms into the stdout but the restapi hides that information.

**Reply by 0xERR0R** [reply]
> Yes. I have a demo project with all rest api endpoints. Its quick and dirty, but it works. [Project](https://github.com/Mozart409/blocky-frontend)
> 
> You can test the "production" version with this [docker-compose](https://github.com/Mozart409/blocky-frontend/blob/main/docker-compose.yml) file. Its included into the project. Be careful. The frontend runs on port [http://localhost:81](http://localhost:81)
> 
> Features
> 
> - enable / disable blocking
> - Refresh Lists
> - Test DNS Queries
> - Responsive Design

Thanks, I'm away for a week, I'll take a look on you demo asap. I think, we will need to introduce some new Rest endpoint later.

**Reply by 0xERR0R** [reply]
Just tested your UI implementation, works like a charm! Do you want to maintain and keep this project in your repository, I would link to your project here (https://0xerr0r.github.io/blocky/installation/#other-installation-types)?

**Reply by Mozart409** [reply]
Thank you. Any things I should change or add? Yes I can keep it there and you can link it. Are you planning to expand the REST API?

**Reply by 0xERR0R** [reply]
If you need a new REST endpoint -> please feel free to create an issue. 

Maybe following things as idea:

- add a screenshot to the readme
- use multi-arch docker build to support arm6/7 etc. -> this would allow to use the web client on Raspberry pi and routers
- button to disable blocking for 5 min
- dark mode ;)

**Reply by Mozart409** [reply]
@0xERR0R Hi, can you teach me how to create multi-arch docker images?

**Reply by kwitsch** [reply]
The build-and-push action you aleady [use](https://github.com/Mozart409/blocky-frontend/blob/main/.github/workflows/docker-publish.yml#L89) supports this feature as it uses buildx internally.
Simply add the platformes you like to bundle together in the [with section](https://github.com/0xERR0R/blocky/blob/development/.github/workflows/release.yml#L60).

**Reply by Mozart409** [reply]
@kwitsch thank you very much.

**Comment by GabeDuarteM** [comment]
@0xERR0R Hey :wave:  I was having [some problems](https://github.com/Mozart409/blocky-frontend/issues/67) running the existing `blocky-frontend`, and since I was also missing a few features, I decided to build a new GUI for Blocky implementing the endpoints available through the API, and from this, BlockyUI was born:

![BlockyUI-Screenshot](https://github.com/GabeDuarteM/blocky-ui/blob/main/docs/BlockyUI-Screenshot.png?raw=true)

It shows you the current state of the DNS server, allows you to enable and disable it (using timers too), allows you to clear the cache and reload the allow/deny lists, and lastly, you can also perform DNS queries with it (and if you use blocky with a DB, you can also view the query log history).

Feel free to check [the webui repo](https://github.com/GabeDuarteM/blocky-ui) out, it has deploy instructions there, and also please let me know if you have any feedbacks or ideas! :smile:

**Reply by 0xERR0R** [reply]
Hi, nice and clean UI, looks very promising. Didn't test it yes, but will try it out. As feedback from the first look, pleass add support for at least arm6 and arm7, so it can run on small edge devices like raspberry pi.

You can also add a link to your project to blocky's documentation.

**Reply by GabeDuarteM** [reply]
Thank you for the early feedback!

About the rpi builds, it should work with raspberry pi 4 and up (I'm using it on my rpi 4). For arm6 and arm7 (I assume for rpi 3 and older), seems like I unfortunately can't build nextjs on those platforms due to lightningcss (a nextjs' dependency) not working on them. I can try to better investigate it further later today to see if theres any workaround

**Reply by GabeDuarteM** [reply]
I asked on their repo and it seems like they don't support it due to [low usage and increased overhead](https://github.com/parcel-bundler/lightningcss/issues/988#issuecomment-2899399207)...

In any case, I managed to do some dark sorcery with docker and built it on arm64 and reused its assets on the other platforms. Seems to work for amd64 and arm64, but unfortunately I can't test it on armv6/v7, so I'm just hoping it also works there :sweat_smile:
